pkg/apis/ml/v1alpha1/storage_spec_backend_s3.go

```diff
@@ -0,0 +1,116 @@
+//
+// DISCLAIMER
+//
+// Copyright 2023 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+
+package v1alpha1
+
+import (
+	"net/url"
+
+	"github.com/arangodb/kube-arangodb/pkg/util/errors"
+)
+
+type ArangoMLStorageSpecBackendS3 struct {
+	// Endpoint specifies the S3 API-compatible endpoint which implements storage
+	// Required
+	Endpoint *string `json:"endpoint"`
+	// BucketName specifies the name of the bucket
+	// Required
+	BucketName *string `json:"bucketName"`
+	// CredentialsSecretName specifies the name of the secret containing AccessKey and SecretKey for S3 API authorization
+	// Required
+	CredentialsSecretName *string `json:"credentialsSecretName"`
+	// AllowInsecure if set to true, the Endpoint certificates won't be checked
+	// +doc/default: false
+	AllowInsecure *bool `json:"allowInsecure,omitempty"`
+	// CASecretName if not empty, the given secret will be used to check the authenticity of Endpoint
+	// The specified `Secret`, must contain the following data fields:
+	// - `ca.crt` PEM encoded public key of the CA certificate
+	// - `ca.key` PEM encoded private key of the CA certificate
+	// +doc/default: ""
+	CASecretName *string `json:"caSecretName,omitempty"`
+	// Region defines the availability zone name. If empty, defaults to 'us-east-1'
+	// +doc/default: ""
+	Region *string `json:"region,omitempty"`
+}
+
+func (s *ArangoMLStorageSpecBackendS3) Validate() error {
+	if s == nil {
+		s = &ArangoMLStorageSpecBackendS3{}
+	}
+
+	if s.GetBucketName() == "" {
```

Comment:
Try to use one execution of shared.WithErrors(shared.PrefixResourceError("s3", s.S3.Validate())) with subfunction for validation - it will give you then full overview about issues in the Spec (not only on the first one)

Reply:
Done